### ISSUE TITLE
Custom property setValue being called multiple times when the value is an object

### ISSUE THREAD

**Issue by silentninja** [schibsted/layout#127]
Current Behaviour:
When setting the value of a custom property as mentioned here https://github.com/schibsted/layout#custom-property-types , setValue is being called multiple times with same value name but If it is a native type setValue is called only once. 

Expected behaviour:
setValue should be called only once when a custom object is used instead of native type.

More details:
It gets called once when the frame size is set to 0 and then thrice with the specified size being set.

**Comment by nicklockwood** [comment]
For non-constant expressions, Layout will often set the value multiple times. There are optimizations in place to avoid setting the value again if it hasn’t changed, but this only works for values that conform to Equatable.

It’s possible you can fix this by implementing Equatable for your custom object. If you have already done this, then this might be a bug, in which case a code sample or unit test that demonstrates the problem would be very helpful.

**Comment by silentninja** [comment]
I tried implementing equatable protocol but the equatable method is not even being called. I tried comparing two objects separately and equatable method gets called and returns the comparison value correctly.

```
// cell.xml
        <UILabel
                height="80"
                left="8"
                top="8"
                obj={customObj}
                width="80" />


//model.swift

public struct CustomData: Codable, Equatable {
    public static func ==(lhs: CustomObj, rhs: CustomObj) -> Bool {
       // Not getting called
        print(lhs.uuid == rhs.uuid)
        print("Custom object is being compared")
        return lhs.uuid == rhs.uuid
    }

    public var uuid: Int?

    public enum CodingKeys: String, CodingKey {
        case uuid
    }
}

//UILabel+TIming.swift

extension UILabel {
    @objc open override class var expressionTypes: [String: RuntimeType] {
        var types = super.expressionTypes
        types["obj"] = RuntimeType(CustomObj.self)
        return types
    }

    @objc open override func setValue(_ value: Any, forExpression name: String) throws {
        switch name {
        case "obj":
            print("Setting obj value") //called multiple times
        default:
            try super.setValue(value, forExpression: name)
        }
    }
}

```

**Comment by nicklockwood** [comment]
@silentninja sorry, I think I was wrong about Equatable being sufficient. Can you try implementing Hashable?

(FYI, if you are using Swift 4.1 or above I think you can omit the implementation of  == and hashValue and Swift will synthesize them automatically.)

**Comment by silentninja** [comment]
When conforming to Hashable setValue method of the extension is not being called when again when the frame height is set( gets called only once when the frame size is 0).  

I noticed that when using tableview, reloading data and sections as soon as the tableview is set reduces the call to 2 times(one when the frame is 0 and when the frame is of specified size).
